src/RecursiveStagesService.php

```diff
@@ -0,0 +1,187 @@
+     * @throws Exception
+     */
+    protected function determineStagesDifferRecursive(DataObject $object): bool
+    {
+        if (!$object->exists()) {
+            // Model hasn't been saved to DB, so we can just bail out as there is nothing to inspect
+            return false;
+        }
+
+        $models = [$object];
+
+        // Compare existing content (perform full ownership traversal)
+        while ($model = array_shift($models)) {
+            if ($this->isModelDirty($model)) {
+                // Model is dirty,
+                // we can return here as there is no need to check the rest of the hierarchy
+                return true;
+            }
+
+            // Discover and add owned objects for inspection
+            $relatedObjects = $this->getOwnedObjects($model);
+            $models = array_merge($models, $relatedObjects);
+        }
+
+        // Compare deleted content,
+        // this wouldn't be covered in hierarchy traversal as deleted models are no longer present
+        $draftIdentifiers = $this->getOwnedIdentifiers($object, Versioned::DRAFT);
+        $liveIdentifiers = $this->getOwnedIdentifiers($object, Versioned::LIVE);
+
+        return $draftIdentifiers !== $liveIdentifiers;
+    }
+
+    /**
+     * Get unique identifiers for all owned objects, so we can easily compare them
+     *
+     * @param DataObject $object
+     * @param string $stage
+     * @return array
+     * @throws Exception
+     */
+    protected function getOwnedIdentifiers(DataObject $object, string $stage): array
+    {
+        $identifiers = Versioned::withVersionedMode(function () use ($object, $stage): array {
+            Versioned::set_stage($stage);
+
+            /** @var DataObject $stagedObject */
+            $stagedObject = DataObject::get_by_id($object->ClassName, $object->ID);
+
+            if ($stagedObject === null) {
+                return [];
+            }
+
+            $models = [$stagedObject];
+            $identifiers = [];
+
+            while ($model = array_shift($models)) {
+                // Compose a unique identifier, so we can easily compare models
+                // Note that we intentionally use base class here, so we can cover the situation where model class changes
+                $identifiers[] = implode('_', [$model->baseClass(), $model->ID]);
+                $relatedObjects = $this->getOwnedObjects($model);
+                $models = array_merge($models, $relatedObjects);
+            }
+
+            return $identifiers;
+        });
+
+        sort($identifiers, SORT_STRING);
+
+        return array_values($identifiers);
+    }
+
+    /**
+     * This lookup will attempt to find "owned" objects
+     * This method uses the 'owns' relation, same as @see RecursivePublishable::publishRecursive()
+     *
+     * @param DataObject|RecursivePublishable $object
+     * @return array
+     * @throws Exception
+     */
+    protected function getOwnedObjects(DataObject $object): array
```

Comment:
```suggestion
    private function getOwnedObjects(DataObject $object): array
```
There's no need for this to be exposed as a protected method at this stage. If someone wants to extend this functionality in the future we can consider their use case at that time, and decide if we want to open this up to be part of our [public API](https://docs.silverstripe.org/en/5/contributing/code/#definition-of-public-api).

Please do this for any other protected methods you've added

Reply:
See my other comment about protected methods.

Reply:
I'll leave these as they are since we're keeping the service.